mapbox_vector_tile/polygon.py

```diff
@@ -43,34 +45,134 @@ def _coords(shape):
     return coords
 
 
+def _drop_degenerate_inners(shape):
+    """
+    Drop degenerate (zero-size) inners from the polygon.
+
+    This is implemented as dropping anything with a size less than 0.5, as the
+    polygon is in integer coordinates and the smallest valid inner would be a
+    triangle with height and width 1.
+    """
+
+    assert shape.geom_type == 'Polygon'
+
+    new_inners = []
+    for inner in shape.interiors:
+        # need to make a polygon of the linearring to get the _filled_ area of
+        # the closed ring.
+        if abs(Polygon(inner).area) >= 0.5:
+            new_inners.append(inner)
+
+    return Polygon(shape.exterior, new_inners)
+
+
+def _contour_to_poly(contour):
+    poly = Polygon(contour)
+    if not poly.is_valid:
+        poly = poly.buffer(0)
+    assert poly.is_valid, \
+        "Contour %r did not make valid polygon %s because %s" \
+        % (contour, poly.wkt, explain_validity(poly))
+    return poly
+
+
+def _polytree_node_to_shapely(node):
+    """
+    Recurses down a Clipper PolyTree, extracting the results as Shapely
+    objects.
+
+    Returns a tuple of (list of polygons, list of children)
+    """
+
+    polygons = []
+    children = []
+    for ch in node.Childs:
+        p, c = _polytree_node_to_shapely(ch)
+        polygons.extend(p)
+        children.extend(c)
+
+    if node.IsHole:
+        # check expectations: a node should be a hole, _or_ return children.
+        # this is because children of holes must be outers, and should be on
+        # the polygons list.
+        assert len(children) == 0
+        if node.Contour:
+            children = [node.Contour]
+        else:
+            children = []
+
+    elif node.Contour:
+        poly = _contour_to_poly(node.Contour)
+        for ch in children:
+            inner = _contour_to_poly(ch)
+            diff = poly.difference(inner)
+            if not diff.is_valid:
+                diff = diff.buffer(0)
+
+            # keep this for when https://trac.osgeo.org/geos/ticket/789 is
+            # resolved.
+            #
+            #  assert diff.is_valid, \
+            #      "Difference of %s and %s did not make valid polygon %s " \
+            #      " because %s" \
+            #      % (poly.wkt, inner.wkt, diff.wkt, explain_validity(diff))
+            #
+            # NOTE: this throws away the inner ring if we can't produce a
+            # valid difference. not ideal, but we'd rather produce something
+            # that's valid than nothing.
+            if diff.is_valid:
+                poly = diff
+
+        assert poly.is_valid
+        polygons.append(poly)
+        children = []
+
+    else:
+        assert len(children) == 0
```

Comment:
Good idea! I've added something in 231011b which hopefully makes it clearer why that's there.